library/core/src/option.rs

```diff
@@ -70,10 +70,23 @@
 //! }
 //! ```
 //!
-//! This usage of [`Option`] to create safe nullable pointers is so
-//! common that Rust does special optimizations to make the
-//! representation of [`Option`]`<`[`Box<T>`]`>` a single pointer. Optional pointers
-//! in Rust are stored as efficiently as any other pointer type.
+//! # Representation
+//!
+//! Rust guarantees to optimize the following types `<T>` such that an
+//! [`Option<T>`] has the same size as `T`:
+//!
+//! * [`Box<T>`]
```

Comment:
This wording is pretty confusing to me, I think because it uses the same name `T` to mean two different things. For example this line is talking about what happens when `T = Box<T>` ... which is never true.

Reply:
I struggled to find a good phrasing here; I agree the current phrasing isn't great.

Might something like:

```rust
//! Rust guarantees that the following types are optimised such that their size
//! is identical whether or not they are surrounded by an `Option`:
```

be clearer?

Reply:
Maybe just use `Box<_>` or `Box<U>`. Naming things is generally a good idea, so I'd not remove the `T` in the introduction.

Reply:
Fixed in https://github.com/rust-lang/rust/pull/75454/commits/68209c3fe4e0f5c3758f18e98efc175af31c2e51.